File: scripts/take_photo_server.py
#!/usr/bin/env python
import rospy
import numpy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from streaming_media.srv import *

base_save_photo_path = "/home/images/"
is_rgb_success = False
is_depth_success = False

def save_image(image, filename, path):
    full_path = base_save_photo_path + path + filename

    if path == 'depth/':
        if not cv2.imwrite(full_path, image.astype(np.uint16)):
            rospy.logerr("Failed to save image.")
            return False
        return True
    
    elif path == 'rgb/':
        if not cv2.imwrite(full_path, image):
            rospy.logerr("Failed to save image.")
            return False
        return True

def take_photo_callback(req):
    global is_rgb_success
    global is_depth_success

    is_rgb_success = is_depth_success = False

    if req.record:  # take photo (a frame)
        sub_rgb = rospy.Subscriber("/usb_cam/image_raw", Image, callback_rgb)
        sub_depth = rospy.Subscriber("/camera/depth/image_raw", Image, callback_depth)

        rospy.sleep(1)

        sub_rgb.unregister()
        sub_depth.unregister()

    return TakePhotoResponse(base_save_video_path if is_rgb_success and is_depth_success else "false")



def callback_rgb(msg):
    global is_rgb_success
    try:
        cv_image = CvBridge().imgmsg_to_cv2(msg, "bgr8")
        rgb_path = "rgb/"
        if save_image(cv_image, "rgb_image.jpg", rgb_path):
            rospy.loginfo("RGB image saved to: %s" % rgb_path)
            is_rgb_success = True
    except CvBridgeError as e:
        rospy.logerr("CvBridge Error: {0}".format(e))

def callback_depth(msg):
    global is_depth_success
    try:
        cv_image = CvBridge().imgmsg_to_cv2(msg, "mono16")
        depth_path = "depth/"
        if save_image(cv_image, "depth_image.png", depth_path):
            rospy.loginfo("Depth image saved to: %s" % depth_path)
            is_depth_success = True
    except CvBridgeError as e:
        rospy.logerr("CvBridge Error: {0}".format(e))

if __name__ == "__main__":
    rospy.init_node("take_photo_server")
    service = rospy.Service("/take_photo", TakePhoto, take_photo_callback)
    rospy.loginfo("Ready to take photos.")
    rospy.spin()

File: src/take_photo_server.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include "streaming_media/TakePhoto.h"  // streaming_media 包名
#include <string>

std::string base_save_photo_path = "/home/images/";

bool saveImage(const cv::Mat& image, const std::string& filename, std::string& path) {
    std::string full_path = base_save_photo_path + path + filename;
    if (!cv::imwrite(full_path, image)) {
        ROS_ERROR("Failed to save image.");
        return false;
    }
    return true;
}

bool takePhotoCallback(streaming_media::TakePhoto::Request  &req, streaming_media::TakePhoto::Response &res) {
    bool is_rgb_success = false, is_depth_success = false;
    if (req.record) {  // take photo(a frame)
        ros::NodeHandle nh;
        image_transport::ImageTransport it(nh);
        image_transport::Subscriber rgb_sub = it.subscribe("/usb_cam/image_raw", 5, [&is_rgb_success](const sensor_msgs::ImageConstPtr& msg) {
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
            std::string rgb_path = "rgb/";
            if (saveImage(cv_ptr->image, "rgb_image.jpg", rgb_path)) {
                ROS_INFO("RGB image saved to: %s", rgb_path.c_str());
                is_rgb_success = true;
            } else ROS_ERROR("Failed to save rgb image.");
        });

        image_transport::Subscriber depth_sub = it.subscribe("/camera/depth/image_raw", 5, [&is_depth_success](const sensor_msgs::ImageConstPtr& msg) {
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "mono16");
            std::string depth_path = "depth/";
            if (saveImage(cv_ptr->image, "depth_image.png", depth_path)) {
                ROS_INFO("Depth image saved to: %s", depth_path.c_str());
                is_depth_success = true;
            } else ROS_ERROR("Failed to save depth image.");
        });

        ros::Duration(10.0).sleep();
        ros::spinOnce();

        if (is_rgb_success && is_depth_success) {
            res.result = base_save_photo_path;  // Return the folder path
        } else res.result = "false";
    }
    return true;
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "take_photo_server");
    ros::NodeHandle nh;

    ros::ServiceServer service = nh.advertiseService("/take_photo", takePhotoCallback);
    ROS_INFO("Ready to take photos.");
    ros::spin();

    return 0;
}

